### PR TITLE
Llm: Initial support of langchain transformer int4 API

### PR DESCRIPTION
**Pull request by rnwang04**
## Description

This PR provides a starting point of langchain transformer int4 API that we support two new llm class（`TransformersLLM` and `TransformersPipelineLLM`) and a new embedding class `TransformersEmbeddings`.

### 1. Why the change?

To provide a starting point of langchain transformer int4 API.

### 2. User API changes

```python
# llm
from bigdl.llm.langchain.llms import TransformersLLM, TransformersPipelineLLM
# text generation pipeline with llama-7b transformer int4 model
llm = TransformersPipelineLLM.from_model_id(
  model_id="decapoda-research/llama-7b-hf",
  task="text-generation",
  model_kwargs={"temperature": 0, "max_length": 64}
)
# chatglm-6b transformer int4 model
llm = TransformersLLM.from_model_id(
  model_id='THUDM/chatglm-6b',
  model_kwargs={"temperature": 0, "max_length": 64, "trust_remote_code": True}
)
llm_chain = LLMChain(prompt=prompt, llm=llm)
output = llm_chain.run(question)

# embedding
from bigdl.llm.langchain.embeddings import TransformersEmbeddings
 embeddings = TransformersEmbeddings.from_model_id(model_id="decapoda-research/llama-7b-hf")
```

### 3. Summary of the change 

- Support `TransformersLLM` to integrate transformer int4 model into langchain llm, based on https://python.langchain.com/docs/modules/model_io/models/llms/how_to/custom_llm, which theoretically can be used for any transformer model
- Support `TransformersPipelineLLM` to integrate transformer pipeline with our optimized transformer int4 model, based on https://python.langchain.com/docs/modules/model_io/models/llms/integrations/huggingface_pipelines, which can't work for chatglm
- Support `TransformersEmbeddings` to integrate transformer int4 model  into langchain embeddings
- Initial example of chat
- Initial example of docqa

### 4. How to test?
- [x] Unit test
- [x] Local test

### Future work
- [ ] stream chat
- [ ] update examples
- [ ] code detail optimization, like accept converted model, pass in model/tokenizer class or flag...

### CODE REVIEW

**Comment by jason-dai** [comment]
Change to `TransformersLLM` and `TransformersPipelineLLM` respectively. Also see https://github.com/analytics-zoo/nano/issues/410

**Comment by rnwang04** [comment]
> Change to `TransformersLLM` and `TransformersPipelineLLM` respectively. Also see [analytics-zoo/nano#410](https://github.com/analytics-zoo/nano/issues/410)

Have change to `TransformersLLM`, `TransformersPipelineLLM` and `TransformersEmbeddings`

File: python/llm/src/bigdl/llm/langchain/llms/bigdltransformersllm.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
+#
```
Comment (jason-dai) [inline python/llm/src/bigdl/llm/langchain/llms/bigdltransformersllm.py:1]: change to transformersllm.py

File: python/llm/src/bigdl/llm/langchain/llms/bigdlpipelinellm.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
+#
```
Comment (jason-dai) [inline python/llm/src/bigdl/llm/langchain/llms/bigdlpipelinellm.py:1]: change to transformerspipelinellm.py

File: python/llm/example/langchain/transformers_int4/chat.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+#
+
+# This would makes sure Python is aware there is more than one sub-package within bigdl,
+# physically located elsewhere.
+# Otherwise there would be module not found error in non-pip's setting as Python would
+# only search the first bigdl package and end up finding only one sub-package.
+
+import argparse
+
+from bigdl.llm.langchain.llms import BigdlTransformersLLM, BigdlPipelineLLM
```
Comment (jason-dai) [inline python/llm/example/langchain/transformers_int4/chat.py:24]: TransformersLLM, TransformersPipelineLLM?

Comment (rnwang04) [reply python/llm/example/langchain/transformers_int4/chat.py]: done

File: python/llm/src/bigdl/llm/langchain/llms/bigdltransformersllm.py (hunk 1 above)
Comment (rnwang04) [reply python/llm/src/bigdl/llm/langchain/llms/bigdltransformersllm.py]: done

**Comment by jason-dai** [comment]
LGTM

**Review by plusbang — APPROVED** [review]
LGTM

**Pull request merged**